### ISSUE TITLE
Enable ExperimentalMemCheck through CTest

### ISSUE THREAD

**Issue by andschwa** [Microsoft/openenclave#761]
Per #731, it is mentioned in the advanced testing guide that `ctest -D ExperimentalMemCheck` can be used to run the tests through Valgrind, however, this currently fails on `master` like this:

```
   Site: acc-22824108
   Build name: Linux-clang++-7
Memory check project /home/andschwa/openenclave/build
Memory checker (MemoryCheckCommand) not set, or cannot find the specified program.
Errors while running CTest
```

It does not appear to be completely setup (or tested since it was first added).

For now, let's remove the `include(CTest)` directive, and the mention of its support in #731.

This issue tracks adding adding (tested) support for it to the build.

**Comment by aserdean** [comment]
I am trying to figure out how the best approach for this one is.

The error that you mentioned (i.e. from my machine):
```
   Site: ubuntu
   Build name: Linux-clang++-7
Create new tag: 20190128-1433 - Experimental
Memory check project /home/ubuntu/openenclave/build
Memory checker (MemoryCheckCommand) not set, or cannot find the specified program.
Errors while running CTest
```
using the following diff:
<details>
ubuntu@ubuntu:~/openenclave/build$ git diff ../CMakeLists.txt
diff --git a/CMakeLists.txt b/CMakeLists.txt
index a6cea47..9055087 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -26,6 +26,7 @@ set(CMAKE_C_COMPILER_NAMES clang-7 cc)
 set(CMAKE_CXX_COMPILER_NAMES clang++-7 c++)

 project("Open Enclave SDK" VERSION ${OE_VERSION} LANGUAGES C CXX ${OE_ASM})
+include (CTest)

 list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake")
</details>

is caused by `MEMORYCHECK_COMMAND ` not being set, due to the lack of `valgrind` (or equivalent)
on the system.

`ExperimentalMemCheck` requires a `DartConfiguration` that can be auto-generated by
importing either `CTest` or `Dart`, as per (https://gitlab.kitware.com/cmake/community/wikis/doc/ctest/Generating-Testing-Files).
We can also generate our configuration for it if needed. I'm guessing we should just use the auto-generated ones but I'm leaving this option here if someone has strong feelings about it.

Should we just add the include of CTest/Dart if `valgrind` is available or should we wrap it under a different target instead of `ExperimentalMemCheck`?

Should we add a step to check for `valgrind` availability and add options to it? (i.e.:  `--leak-check=full`; suppression files; etc?)

**Comment by andschwa** [comment]
I think just import CTest if valgrind is available, I'm fine with the `ExperimentalMemCheck` target.

Definitely should check for valgrind before enabling anything here.

I have no strong opinions on this, if it works fairly easily, that would be great 😄

**Comment by aserdean** [comment]
Thanks a lot for the quick response!

As an update.

Using the following diff on master:
<details>
diff --git a/CMakeLists.txt b/CMakeLists.txt
index a6cea47..6ba8f54 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -130,6 +130,16 @@ endif ()

 option(ADD_WINDOWS_ENCLAVE_TESTS "Build Windows enclave tests" OFF)

+# add valgrind
+find_program(VALGRIND "valgrind")
+if (VALGRIND)
+  include(Dart)
+  message(STATUS "ExperimentalMemCheck can be used to run tests under valgrind")
+else ()
+  message(STATUS "Valgrind not found")
+endif ()
+
+
 # Configure testing
 enable_testing()
 include(add_enclave_test)
</details>

with `valgrind` installed and running the tests with the following command:
`OE_SIMULATION=1 ctest -D ExperimentalMemCheck -j --timeout=30 -VV >output.txt`
I get the following failures:
```
The following tests FAILED:
         12 - tests/crypto/enclave (SEGFAULT)
         13 - tests/crypto/host (Failed)
         40 - tests/thread_local (Child killed)
         41 - tests/thread_local_exported (SEGFAULT)
         49 - tests/libc (Child aborted)
         94 - tests/libcxxtest-std_thread_thread.mutex_thread.lock_thread.lock.shared_thread.lock.shared.locking_lock.pass (Child killed)
         96 - tests/libcxxtest-libcxx_utilities_memory_util.smartptr_race_condition.pass (Timeout)
        112 - tests/mbed/selftest (SEGFAULT)
        128 - tests/mbed/entropy (SEGFAULT)
        151 - tests/mbed/rsa (SEGFAULT)
```

The full output (including the` valgrind` analysis can be found here:
[output.txt](https://github.com/Microsoft/openenclave/files/2804245/output.txt)

Should we deal with them before we address this issue or after?

**Comment by aserdean** [comment]
I sent a tentative at: https://github.com/Microsoft/openenclave/pull/1399

**Comment by andschwa** [comment]
> Should we deal with them before we address this issue or after?

Nah, definitely after. The only thing that matters to me here is it running them under Valgrind, just a useful tool.

**Comment by andschwa** [comment]
Closed via #1399.
